openebs/maya
# fix(pvc deletion, webhook): add validation prior to deleting a pvc

### Why we need it
When PVCs are deleted, it deletes various openebs resources associated with the PV. There are cases where this volume might have clones. Clones are dependent on existence of this volume _(referred to as source volume w.r.t a clone)_. There are certain validations later in the deletion lifecycle, however it still leaves this PV in an orphaned state. In other words, some of the resources get deleted while some stick around due to validations. This PR solves the problem of partial deletion by introducing web hook which enable injecting validation(s) right at the start of the lifecycle.

### What this PR does

This PR is a implementation of _kubernetes validation admission webhook_. It will be deployed in kubernetes cluster as a deployment.  It validates each `PVC` delete operation before proceeding with delete. 

Following are the checks of this validation:
- Clone Volume Exists:
  - if clone exists for a PVC that is being deleted, validation will fail & hence this delete operation fails.
  - fixes #openebs/openebs/issues/2350
  - fixes #openebs/openebs/issues/2316

- Volume with/without snapshot (No Clone):
  - If volume with/without snapshot but no clone(s), is being deleted then PVC deletion will succeed.

### How to test deploy Webhook and its configuration:

To deploy the server, we’ll need to create a service and a deployment in our Kubernetes cluster. . It’s pretty straightforward, except one thing, which is the server’s TLS configuration. 
The required yamls are the part of openebs-operator which includes below mentioned configurations. follow the below steps to deploy and test.

Note: Required yamls has been added as part of PR openebs/openebs/pull/2404

### How to test changes:
- Create k8s secret including cert/key pair using `manual-webhook-certs.yaml` (TLS certificate signed by the Kubernetes CA to secure the  communication between the webhook server and apiserver).	
- Create webhook deployment and service using `webhook-deployment.yaml`  (make sure its in running state).
- Create webhook confiuration using `validatingwebhook-ca-bundle.yaml` which specifies what resources and what actions the webhook server handles
- Create clone volume and try to delete the source PVC using `kubectl delete pvc <pvc-name>` .


```bash
 $ kubectl get pods -n openebs
NAME                                                              READY   STATUS    RESTARTS   AGE
maya-apiserver-6d7b758446-vdmxq                                   1/1     Running   0          3h47m
openebs-admission-webhook-67b7fd5886-c8qsz                        1/1     Running   0          19s
openebs-ndm-vmt8g                                                 1/1     Running   0          3h47m
openebs-provisioner-7d594494dd-px5bh                              1/1     Running   0          3h47m
openebs-snapshot-operator-5d58977fcf-2pd8r                        2/2     Running   0          3h47m
pvc-2d1f4f43-343d-11e9-a9c2-54e1ad0c1ccc-target-56769c68ccchvxr   3/3     Running   0          3h13m
pvc-d2b8837b-343c-11e9-a9c2-54e1ad0c1ccc-target-7ff8f9d78756wpg   3/3     Running   0          3h16m
sparse-claim-auto-ev8s-5f6cf5bdc8-sjjwh                           2/2     Running   0          3h45m
```

```bash
$ kubectl get pvc
NAME                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                AGE
cstor-vol1-1r-claim         Bound    pvc-d2b8837b-343c-11e9-a9c2-54e1ad0c1ccc   4G         RWO            openebs-cstor-sparse-auto   32m
demo-snap-vol-claim-cstor   Bound    pvc-2d1f4f43-343d-11e9-a9c2-54e1ad0c1ccc   4G         RWO            openebs-snapshot-promoter   30m
```

```sh
$ kubectl delete pvc cstor-vol1-1r-claim
Error from server (PVC has cloned volume): admission webhook "admission-webhook.openebs.io" denied the request: pvc "cstor-vol1-1r-claim" has one or more cloned volume(s), count is '1'
```

## Upgrade Steps:
1. We have to only patch the  required `openebs.io/source-volume:<pv-name>` labels in all the Cloned CstorVolumes resources

## Note : 
Tested the changes even creating jiva volumes and works as expected.

**Checklist:**
- [x] Fixes #<issue number>
- [x] Labelled this PR & related issue with `documentation` tag
- [x] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [x] Labelled this PR & related issue with `requires-upgrade` tag
- [x] PR messages has upgrade related information
- [x] Commit has unit tests
- [ ] Commit has integration tests
Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>
